File: packages/sdk/src/domains/conversations.ts
/**
 * The conversations domain: list the conversations this account is in, create new ones, and run the
 * group lifecycle around them.
 *
 * Direct chats are the thin case: {@link create} is *idempotent* on its member set, so creating "the
 * conversation with Alice" twice returns the same conversation rather than a second one — the server
 * derives a deterministic id from the sorted member ids, which is what lets a client call {@link
 * create} freely whenever it needs the conversation to exist before sending.
 *
 * # The group lifecycle
 *
 * A group has two founders — the creator and the first person they named — and the founders are the
 * group's memory of who built it. {@link invite} is every member's right; {@link mute}, {@link kick},
 * and {@link rename} are the founders'; {@link voteKick} is the members' own recourse, carrying at
 * half the roster rounded up. Leaving ({@link leave}) is nobody's to gate: when the last founder
 * walks out, the longest-standing member inherits the role, so a group never reaches a state where
 * nobody can rename it or answer a report.
 *
 * Three live streams follow from {@link start}: {@link onMember} for joins, departures, and removals
 * — rotate sender keys on every one of these, exactly as for a room — {@link onVote} for a running
 * kick tally, and {@link onState} for coalesced metadata deltas such as a rename. All three arrive
 * only on a conversation the client has subscribed to through {@link
 * MigoClient.watchConversation}.
 *
 * The summaries this returns carry an {@link EncryptionMode} the UI may display, but that field is a
 * claim about transport, not a substitute for the end-to-end guarantee: content is sealed by the
 * crypto layers regardless of what mode a summary advertises.
 */

import type { Id } from '@migo/wire';
import {
  OP,
  ConversationKind,
  encodeConversationListRequest,
  decodeConversationListResponse,
  encodeConversationCreateRequest,
  decodeConversationSummary,
  encodeConversationInviteRequest,
  encodeConversationLeaveRequest,
  encodeConversationRosterRequest,
  decodeConversationRosterResponse,
  encodeConversationMuteRequest,
  encodeConversationKickRequest,
  encodeConversationVoteKickRequest,
  decodeConversationVoteKickResponse,
  encodeConversationUpdateRequest,
  decodeConversationMemberEvent,
  decodeConversationVoteEvent,
  decodeConversationStateEvent,
  decodeAcknowledged,
} from '@migo/protocol';
import type {
  ConversationListRequest,
  ConversationListResponse,
  ConversationCreateRequest,
  ConversationSummary,
  ConversationInviteRequest,
  ConversationLeaveRequest,
  ConversationRosterRequest,
  ConversationRosterResponse,
  ConversationRosterEntry,
  ConversationMuteRequest,
  ConversationKickRequest,
  ConversationVoteKickRequest,
  ConversationVoteKickResponse,
  ConversationUpdateRequest,
  ConversationMemberEvent,
  ConversationVoteEvent,
  ConversationStateEvent,
  Acknowledged,
} from '@migo/protocol';

import { ListenerSet } from './listeners.js';
import type { Listener } from './listeners.js';
import type { EventErrorHandler, Rpc } from './rpc.js';

/** Optional creation parameters beyond the member set. */
export interface CreateConversationOptions {
  /** A title, meaningful for group conversations; ignored by the server for direct ones. */
  title?: string;
}

/**
 * List, create, and run the lifecycle of conversations.
 *
 * One instance per client. {@link start} begins delivering the three group event streams; the
 * request/response methods work independently of it.
 */
export class ConversationsDomain {
  readonly #rpc: Rpc;
  readonly #memberListeners: ListenerSet<ConversationMemberEvent>;
  readonly #voteListeners: ListenerSet<ConversationVoteEvent>;
  readonly #stateListeners: ListenerSet<ConversationStateEvent>;
  #unsubscribes: Array<() => void> = [];

  constructor(rpc: Rpc, onEventError?: EventErrorHandler) {
    this.#rpc = rpc;
    this.#memberListeners = new ListenerSet(OP.CONVERSATION_MEMBER_EVENT, onEventError);
    this.#voteListeners = new ListenerSet(OP.CONVERSATION_VOTE_EVENT, onEventError);
    this.#stateListeners = new ListenerSet(OP.CONVERSATION_STATE_EVENT, onEventError);
  }

  /** Begins delivering group member, vote, and state events. Idempotent. */
  start(): void {
    if (this.#unsubscribes.length > 0) {
      return;
    }
    this.#unsubscribes.push(
      this.#rpc.on(OP.CONVERSATION_MEMBER_EVENT, decodeConversationMemberEvent, (event) =>
        this.#memberListeners.deliver(event),
      ),
      this.#rpc.on(OP.CONVERSATION_VOTE_EVENT, decodeConversationVoteEvent, (event) =>
        this.#voteListeners.deliver(event),
      ),
      this.#rpc.on(OP.CONVERSATION_STATE_EVENT, decodeConversationStateEvent, (event) =>
        this.#stateListeners.deliver(event),
      ),
    );
  }

  /** Stops delivering conversation events. Registered handlers are kept for a later {@link start}. */
  stop(): void {
    for (const unsubscribe of this.#unsubscribes) {
      unsubscribe();
    }
    this.#unsubscribes = [];
  }

  /**
   * Registers a handler for group membership movement: joins, departures, and removals. Returns an
   * unsubscribe function.
   *
   * Membership churn is a crypto event before it is a UI one — rotate the conversation's sender key
   * on every one of these, exactly as for a room, so a removed member cannot read what is sent next.
   */
  onMember(handler: Listener<ConversationMemberEvent>): () => void {
    return this.#memberListeners.add(handler);
  }

  /**
   * Registers a handler for group kick-vote tallies. Returns an unsubscribe function.
   *
   * Each event names the target and the count so far against the threshold; `closed` marks the vote's
   * end (it passed, expired, or the target walked out). The stream is coalesced per conversation, so a
   * handler holds one tally per target and replaces it as counts arrive rather than accumulating.
   */
  onVote(handler: Listener<ConversationVoteEvent>): () => void {
    return this.#voteListeners.add(handler);
  }

  /**
   * Registers a handler for coalesced group metadata deltas. Returns an unsubscribe function.
   *
   * Each event carries only the fields that changed — today, a `title` when a founder renames the
   * group. Apply them onto held state rather than replacing it.
   */
  onState(handler: Listener<ConversationStateEvent>): () => void {
    return this.#stateListeners.add(handler);
  }

  /**
   * Lists the account's conversations, newest activity first.
   *
   * Pass the {@link ConversationListResponse.nextCursor} from a previous page as `cursor` to fetch the
   * next; a response with no `nextCursor` is the last page. `limit` bounds one page.
   */
  async list(limit: number, cursor?: string): Promise<ConversationListResponse> {
    const request: ConversationListRequest = { limit };
    if (cursor !== undefined) {
      request.cursor = cursor;
    }
    return this.#rpc.call(
      OP.CONVERSATION_LIST,
      encodeConversationListRequest,
      decodeConversationListResponse,
      request,
    );
  }

  /**
   * Creates a conversation, or returns the existing one for a direct chat.
   *
   * `members` is the *other* participants; the server adds the caller. For a {@link
   * ConversationKind.Direct} chat this is idempotent — the same two members always resolve to the same
   * conversation — so callers may use it as "ensure the conversation exists" before a first send.
   */
  async create(
    kind: ConversationKind,
    members: Id[],
    options: CreateConversationOptions = {},
  ): Promise<ConversationSummary> {
    const request: ConversationCreateRequest = { kind, members };
    if (options.title !== undefined) {
      request.title = options.title;
    }
    return this.#rpc.call(
      OP.CONVERSATION_CREATE,
      encodeConversationCreateRequest,
      decodeConversationSummary,
      request,
    );
  }

  /**
   * Adds members to a group, resolving with the group's summary as it now stands.
   *
   * Any current member may invite; the new seats arrive as plain members. `members` holds account ids
   * — to invite by username, resolve the name to an id first (an account search), then pass the id
   * here. Already-seated members and the caller are quietly skipped, and each person who actually
   * landed is announced to the group on the {@link onMember} stream, so the roster stays true without
   * a refetch. Not the group's founders, so the group always has at least one.
   */
  async invite(conversationId: Id, members: Id[]): Promise<ConversationSummary> {
    const request: ConversationInviteRequest = { conversationId, members };
    return this.#rpc.call(
      OP.CONVERSATION_INVITE,
      encodeConversationInviteRequest,
      decodeConversationSummary,
      request,
    );
  }

  /**
   * Leaves a group. Nobody's permission is asked — leaving is a right, not a request.
   *
   * After leaving, forget the conversation's crypto state through the messaging domain, exactly as for
   * a room: the group rotates its sender key on the departure, and the local receiver state is no
   * longer useful. When the last founder leaves, the longest-standing member silently inherits the
   * role, so the group never ends up with nobody able to rename it or answer a report.
   */
  async leave(conversationId: Id): Promise<Acknowledged> {
    const request: ConversationLeaveRequest = { conversationId };
    return this.#rpc.call(
      OP.CONVERSATION_LEAVE,
      encodeConversationLeaveRequest,
      decodeAcknowledged,
      request,
    );
  }

  /**
   * Reads a group's roster: active members first by join time, then the departed.
   *
   * Each {@link ConversationRosterEntry} carries the account, its {@link ConversationRole}, when it
   * joined, and any group mute still running against them. Departed members carry a `leftAt`, which is
   * how a UI renders "was here" without pretending history is deletable. This is a snapshot for
   * rendering, not a mirror to maintain — live movement arrives on {@link onMember}.
   */
  async getRoster(conversationId: Id): Promise<ConversationRosterEntry[]> {
    const request: ConversationRosterRequest = { conversationId };
    const response: ConversationRosterResponse = await this.#rpc.call(
      OP.CONVERSATION_ROSTER,
      encodeConversationRosterRequest,
      decodeConversationRosterResponse,
      request,
    );
    return response.entries;
  }

  /**
   * Mutes or unmutes one member of a group — a founder's action, not a vote.
   *
   * While the mute runs, the target cannot send to the group; they keep every other right, including
   * the vote. Omit `until` to lift a mute early; pass a future epoch-milliseconds timestamp to set
   * one. Founders are beyond each other's reach, and neither the caller nor a founder may be the
   * target. There is no event for this — the roster is the record — so a client that needs to show
   * the change refetches the roster.
   */
  async mute(conversationId: Id, targetId: Id, until?: number): Promise<Acknowledged> {
    const request: ConversationMuteRequest = { conversationId, targetId };
    if (until !== undefined) {
      request.until = until;
    }
    return this.#rpc.call(
      OP.CONVERSATION_MUTE,
      encodeConversationMuteRequest,
      decodeAcknowledged,
      request,
    );
  }

  /**
   * Removes a member outright, no vote — a founder's action.
   *
   * The other founder is beyond this reach: a group built by two cannot be halved by one of them. The
   * removal is announced to the group on the {@link onMember} stream, and the group rotates its sender
   * key, so rotate local crypto state exactly as for any membership change.
   */
  async kick(conversationId: Id, targetId: Id): Promise<Acknowledged> {
    const request: ConversationKickRequest = { conversationId, targetId };
    return this.#rpc.call(
      OP.CONVERSATION_KICK,
      encodeConversationKickRequest,
      decodeAcknowledged,
      request,
    );
  }

  /**
   * Casts a voice to kick a member by vote, resolving with the tally after this voice landed.
   *
   * The vote is the members' own recourse, needing no rank: the first call opens the vote, each
   * further member's call adds to it, and when {@link ConversationVoteKickResponse.votes} reaches
   * `needed` — half the group, rounded up — the kick lands and `open` turns false. A caller's
   * repeated voice is idempotent, a muted member still votes, and founders are immune (a group's
   * founders are answerable to nobody but each other). One vote may run per group at a time, and a
   * vote nobody finishes expires after a minute. The same tally is broadcast on the {@link onVote}
   * stream, so a member who voted and one who only watched converge on the same count.
   */
  async voteKick(conversationId: Id, targetId: Id): Promise<ConversationVoteKickResponse> {
    const request: ConversationVoteKickRequest = { conversationId, targetId };
    return this.#rpc.call(
      OP.CONVERSATION_VOTE_KICK,
      encodeConversationVoteKickRequest,
      decodeConversationVoteKickResponse,
      request,
    );
  }

  /**
   * Renames a group, resolving with the summary carrying the new title.
   *
   * A founder's action. The new title travels to every member on the {@link onState} stream as a
   * coalesced delta, so a client applies it onto the summary it already holds. Direct conversations
   * carry no title to change.
   */
  async rename(conversationId: Id, title: string): Promise<ConversationSummary> {
    const request: ConversationUpdateRequest = { conversationId, title };
    return this.#rpc.call(
      OP.CONVERSATION_UPDATE,
      encodeConversationUpdateRequest,
      decodeConversationSummary,
      request,
    );
  }
}

File: clients/android/core/src/main/kotlin/com/migo/core/domain/Conversations.kt
package com.migo.core.domain

import com.migo.core.protocol.Acknowledged
import com.migo.core.protocol.ConversationCreateRequest
import com.migo.core.protocol.ConversationInviteRequest
import com.migo.core.protocol.ConversationKickRequest
import com.migo.core.protocol.ConversationKind
import com.migo.core.protocol.ConversationLeaveRequest
import com.migo.core.protocol.ConversationListRequest
import com.migo.core.protocol.ConversationListResponse
import com.migo.core.protocol.ConversationMemberEvent
import com.migo.core.protocol.ConversationMuteRequest
import com.migo.core.protocol.ConversationRole
import com.migo.core.protocol.ConversationRosterEntry
import com.migo.core.protocol.ConversationRosterRequest
import com.migo.core.protocol.ConversationRosterResponse
import com.migo.core.protocol.ConversationStateEvent
import com.migo.core.protocol.ConversationSummary
import com.migo.core.protocol.ConversationUpdateRequest
import com.migo.core.protocol.ConversationVoteEvent
import com.migo.core.protocol.ConversationVoteKickRequest
import com.migo.core.protocol.ConversationVoteKickResponse
import com.migo.core.protocol.Op
import com.migo.core.wire.Id

/**
 * Listing the conversations this account is in, creating new ones, and running the group lifecycle
 * around them.
 *
 * A port of `packages/sdk/src/domains/conversations.ts`. Requests and responses only for the
 * lifecycle calls; the three group event streams are delivered through the [start]-driven listener
 * sets, exactly as [RoomsDomain] does it.
 *
 * # The property worth knowing
 *
 * [create] is idempotent for a [ConversationKind.Direct] chat. The server derives the conversation
 * id deterministically from the sorted member ids, so "the conversation with Alice" resolves to the
 * same conversation every time it is asked for. That is what makes it safe to call before a first
 * send rather than tracking whether one has been created, and it is why there is no
 * `ensureConversation` helper here: create already is one.
 *
 * # The group lifecycle
 *
 * A group has two founders -- the creator and the first person they named -- and the founders are
 * the group's memory of who built it. [invite] is every member's right; [mute], [kick], and
 * [rename] are the founders'; [voteKick] is the members' own recourse, carrying at half the roster
 * rounded up. Leaving ([leave]) is nobody's to gate: when the last founder walks out, the
 * longest-standing member inherits the role, so a group never reaches a state where nobody can
 * rename it or answer a report.
 *
 * Three live streams follow from [start]: [onMember] for joins, departures, and removals -- rotate
 * sender keys on every one of these, exactly as for a room -- [onVote] for a running kick tally,
 * and [onState] for coalesced metadata deltas such as a rename. All three arrive only on a
 * conversation the client has subscribed to through the client's `watchConversation`.
 *
 * # What the encryption mode in a summary does and does not say
 *
 * [ConversationSummary.encryption] is a claim about the conversation's transport policy, and a client
 * may show it. It is not the end-to-end guarantee: content is sealed by the crypto layers on the way
 * out regardless of what a summary advertises, so a summary that came back with an unexpected mode
 * cannot cause plaintext to be sent. Treating the field as the source of truth for "is this
 * encrypted" would put a security decision in the hands of the party the encryption protects
 * against.
 */
class ConversationsDomain(
    private val rpc: Rpc,
    onEventError: EventErrorHandler? = null,
) {
    private val memberListeners =
        ListenerSet<ConversationMemberEvent>(Op.CONVERSATION_MEMBER_EVENT, onEventError)
    private val voteListeners =
        ListenerSet<ConversationVoteEvent>(Op.CONVERSATION_VOTE_EVENT, onEventError)
    private val stateListeners =
        ListenerSet<ConversationStateEvent>(Op.CONVERSATION_STATE_EVENT, onEventError)

    @Volatile
    private var subscriptions: List<Subscription>? = null

    /**
     * Begins delivering the group event streams to registered handlers. Idempotent.
     *
     * All three are subscribed together, for the same reason [RoomsDomain.start] takes the room's
     * three as one: the server sends them whether or not a handler asked, and an opcode reaching
     * [Rpc] with no subscriber is a warning rather than a silent discard.
     */
    fun start() {
        if (subscriptions != null) return
        subscriptions = listOf(
            rpc.on(Op.CONVERSATION_MEMBER_EVENT, { r -> ConversationMemberEvent.decode(r) }) { event, _ ->
                memberListeners.deliver(event)
            },
            rpc.on(Op.CONVERSATION_VOTE_EVENT, { r -> ConversationVoteEvent.decode(r) }) { event, _ ->
                voteListeners.deliver(event)
            },
            rpc.on(Op.CONVERSATION_STATE_EVENT, { r -> ConversationStateEvent.decode(r) }) { event, _ ->
                stateListeners.deliver(event)
            },
        )
    }

    /** Stops delivery of all three streams. Registered handlers are kept for a later [start]. */
    fun stop() {
        val live = subscriptions ?: return
        subscriptions = null
        for (subscription in live) subscription.cancel()
    }

    /**
     * Registers a handler for group membership movement: joins, departures, and removals.
     *
     * Membership churn is a crypto event before it is a UI one -- rotate the conversation's sender
     * key on every one of these, exactly as for a room, so a removed member cannot read what is
     * sent next.
     */
    fun onMember(listener: Listener<ConversationMemberEvent>): Subscription = memberListeners.add(listener)

    /**
     * Registers a handler for group kick-vote tallies.
     *
     * Each event names the target and the count so far against the threshold; [ConversationVoteEvent.closed]
     * marks the vote's end (it passed, expired, or the target walked out). The stream is coalesced
     * per conversation, so a handler holds one tally per target and replaces it as counts arrive
     * rather than accumulating.
     */
    fun onVote(listener: Listener<ConversationVoteEvent>): Subscription = voteListeners.add(listener)

    /**
     * Registers a handler for coalesced group metadata deltas.
     *
     * Each event carries only the fields that changed -- today, a title when a founder renames the
     * group. Apply them onto held state rather than replacing it.
     */
    fun onState(listener: Listener<ConversationStateEvent>): Subscription = stateListeners.add(listener)

    /**
     * Lists the account's conversations, most recent activity first.
     *
     * Pass the [ConversationListResponse.nextCursor] of a previous page as [cursor] for the next one;
     * a response without a cursor is the last page. [limit] bounds one page, and the server may
     * return fewer.
     */
    suspend fun list(limit: Long, cursor: String? = null): ConversationListResponse {
        val request = ConversationListRequest(limit, cursor)
        return rpc.call(
            Op.CONVERSATION_LIST,
            { w -> request.encode(w) },
            { r -> ConversationListResponse.decode(r) },
        )
    }

    /**
     * Creates a conversation, or returns the existing one for a direct chat.
     *
     * [members] is the *other* participants: the server adds the caller, and a client that included
     * itself would be asking for a two-person conversation with one person in it. [title] is
     * meaningful for a group and ignored for a direct chat, where the name shown is the other
     * person's. A group needs at least one named member besides the caller -- the two of them are
     * the group's founders -- so an empty member list is refused by the server, not padded here.
     */
    suspend fun create(
        kind: ConversationKind,
        members: List<Id>,
        title: String? = null,
    ): ConversationSummary {
        val request = ConversationCreateRequest(kind, members, title)
        return rpc.call(
            Op.CONVERSATION_CREATE,
            { w -> request.encode(w) },
            { r -> ConversationSummary.decode(r) },
        )
    }

    /**
     * Adds members to a group, resolving with the group's summary as it now stands.
     *
     * Any current member may invite; the new seats arrive as plain members. Already-seated members
     * and the caller are quietly skipped, and each person who actually landed is announced to the
     * group on the [onMember] stream, so the roster stays true without a refetch.
     */
    suspend fun invite(conversationId: Id, members: List<Id>): ConversationSummary {
        val request = ConversationInviteRequest(conversationId, members)
        return rpc.call(
            Op.CONVERSATION_INVITE,
            { w -> request.encode(w) },
            { r -> ConversationSummary.decode(r) },
        )
    }

    /**
     * Leaves a group. Nobody's permission is asked -- leaving is a right, not a request.
     *
     * After leaving, forget the conversation's crypto state through the messaging domain, exactly as
     * for a room: the group rotates its sender key on the departure, and the local receiver state is
     * no longer useful. When the last founder leaves, the longest-standing member silently inherits
     * the role, so the group never ends up with nobody able to rename it or answer a report.
     */
    suspend fun leave(conversationId: Id): Acknowledged {
        val request = ConversationLeaveRequest(conversationId)
        return rpc.call(
            Op.CONVERSATION_LEAVE,
            { w -> request.encode(w) },
            { r -> Acknowledged.decode(r) },
        )
    }

    /**
     * Reads a group's roster: active members first by join time, then the departed.
     *
     * The whole membership, where a conversation-list row's members field is a capped preview --
     * which is why the client's sender-key audience is chosen from this answer and never from a
     * list row. A departure carries a non-null [ConversationRosterEntry.leftAt]; the caller that
     * wants "who is in the group now" filters on it.
     */
    suspend fun getRoster(conversationId: Id): List<ConversationRosterEntry> {
        val request = ConversationRosterRequest(conversationId)
        val response = rpc.call(
            Op.CONVERSATION_ROSTER,
            { w -> request.encode(w) },
            { r -> ConversationRosterResponse.decode(r) },
        )
        return response.entries
    }

    /**
     * Mutes or unmutes one member of a group -- a founder's action, not a vote.
     *
     * While the mute runs, the target cannot send to the group; they keep every other right,
     * including the vote. Omit [until] to lift a mute early; pass a future epoch-milliseconds
     * timestamp to set one. Founders are beyond each other's reach, and neither the caller nor a
     * founder may be the target. There is no event for this -- the roster is the record -- so a
     * client that needs to show the change refetches the roster.
     */
    suspend fun mute(conversationId: Id, targetId: Id, until: Long? = null): Acknowledged {
        val request = ConversationMuteRequest(conversationId, targetId, until)
        return rpc.call(
            Op.CONVERSATION_MUTE,
            { w -> request.encode(w) },
            { r -> Acknowledged.decode(r) },
        )
    }

    /**
     * Removes a member outright, no vote -- a founder's action.
     *
     * The other founder is beyond this reach: a group built by two cannot be halved by one of them.
     * The removal is announced to the group on the [onMember] stream, and the group rotates its
     * sender key, so rotate local crypto state exactly as for any membership change.
     */
    suspend fun kick(conversationId: Id, targetId: Id): Acknowledged {
        val request = ConversationKickRequest(conversationId, targetId)
        return rpc.call(
            Op.CONVERSATION_KICK,
            { w -> request.encode(w) },
            { r -> Acknowledged.decode(r) },
        )
    }

    /**
     * Casts a voice to kick a member by vote, resolving with the tally after this voice landed.
     *
     * The vote is the members' own recourse, needing no rank: the first call opens the vote, each
     * further member's call adds to it, and when [ConversationVoteKickResponse.votes] reaches
     * [ConversationVoteKickResponse.needed] -- half the group, rounded up -- the kick lands and
     * [ConversationVoteKickResponse.open] turns false. A caller's repeated voice is idempotent, a
     * muted member still votes, and founders are immune. One vote may run per group at a time, and
     * a vote nobody finishes expires after a minute. The same tally is broadcast on the [onVote]
     * stream, so a member who voted and one who only watched converge on the same count.
     */
    suspend fun voteKick(conversationId: Id, targetId: Id): ConversationVoteKickResponse {
        val request = ConversationVoteKickRequest(conversationId, targetId)
        return rpc.call(
            Op.CONVERSATION_VOTE_KICK,
            { w -> request.encode(w) },
            { r -> ConversationVoteKickResponse.decode(r) },
        )
    }

    /**
     * Renames a group, resolving with the summary carrying the new title.
     *
     * A founder's action. The new title travels to every member on the [onState] stream as a
     * coalesced delta, so a client applies it onto the summary it already holds. Direct
     * conversations carry no title to change.
     */
    suspend fun rename(conversationId: Id, title: String): ConversationSummary {
        val request = ConversationUpdateRequest(conversationId, title)
        return rpc.call(
            Op.CONVERSATION_UPDATE,
            { w -> request.encode(w) },
            { r -> ConversationSummary.decode(r) },
        )
    }
}

/**
 * Whether the founder controls (mute, kick) belong on a member's row for this viewer.
 *
 * Pure, so a test can pin it. The viewer must be a founder, the target must not be -- the two
 * builders are beyond each other's reach -- and nobody acts on their own row.
 */
fun canFounderAct(viewerRole: ConversationRole, targetRole: ConversationRole, isSelf: Boolean): Boolean =
    !isSelf && viewerRole == ConversationRole.Founder && targetRole != ConversationRole.Founder

/**
 * Whether the "Vote kick" control belongs on a member's row.
 *
 * Pure, so a test can pin it. The vote is every member's own recourse -- but never against
 * yourself, and never against a founder, whom a show of hands cannot unseat.
 */
fun canVoteKickGroup(targetRole: ConversationRole, isSelf: Boolean): Boolean =
    !isSelf && targetRole != ConversationRole.Founder
